app/models/rule.rb

```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class Rule < ApplicationRecord
+  before_validation :error_if_locked, on: :update
+
+  # Allow an authorized user to unlock a rule
+  def self.unlock(rule, user)
+    raise(RuleLockedError, rule.id) unless user.can_unlock?(rule)
```

Comment:
1. Where will privilege checking for locking a rule be done?
2. Is the idea to have an `update` and a `unlock` controller action in the RuleController?

Reply:
Could this be more elegant and set up for authorization in future actions if we remove this static `self.unlock` method and write the manage lock controller action now with pundit?

https://github.com/varvet/pundit

```ruby
class RulesController < ApplicationController
  def update
    authorize @rule
    # TODO
  end

  def manage_lock
    authorize @rule
    rule.locked = manage_lock_params[:rule][:locked]
    rule.save(validate: false)
  end

  private

  def update_params
    params.require(:rule).permit(:todo)
  end

  def manage_lock_params
    params.require(:rule).permit(:locked)
  end
end

class RulePolicy
  attr_reader :user, :rule

  def initialize(user, rule)
    @user = user
    @rule = rule
  end

  def update?
    # TODO
    true
  end

  def manage_lock?
    user.can_unlock?(rule)
  end
end
```

Reply:
Yes we would want a controller action for lock and update

Reply:
What is the reasoning for separating lock/unlock into two separate controller actions?